### ISSUE TITLE
Deprecate creating/realizing tasks after the task graph has been realized

### ISSUE THREAD

**Issue by big-guy** [gradle/gradle-native#868]
This is almost always unintentional and can cause NPE/concurrency problems when combined with lazy tasks.

If we deprecate this, we can see how prevalent this is in practice and then later switch this into an error.  This may also expose more unsafe access that's coincidentally working

**Comment by ghale** [comment]
Just some thoughts on this topic:

I think the primary cause of this is where tasks are required because of the metadata associated with the task object and pretty much never because an actual task object is needed.  The classic example of this is accessing an artifact that is backed by an archive task.  In cases where we don't need the actual artifact built, but only need the metadata about the artifact such as IDE generation, this ends up getting calculated at task execution time (and ultimately leads to task realization).  But this could happen with any deferred `Provider` in the project model.  The point is, this is a product of general `Provider` usage/realization and not something specific to tasks.  

I think a couple fundamental questions come out of this: 

- Is task realization something different from any other sort of project model mutation?  
- If not, then are we implying that project model mutation should never happen at execution time?

It seems like if the answer to the questions above is no/yes, then we need some sort of point where referenced providers are realized before execution starts (which seems like an approach divergent from our current course).  If the answer is no/no, then what we're doing now (project mutation guard) is probably a sane approach but we might want to build something into providers that warns when they are realized without holding the appropriate project lock.

If the answer to the first question above is yes, then I'm not sure what we would need - we'd have to define what makes tasks different first.

**Comment by big-guy** [comment]
> Is task realization something different from any other sort of project model mutation?

No, I don't think it's special beyond that we know it doesn't make sense to add/remove Tasks once we've decided what we're going to execute and tasks are, unfortunately, a carrier for domain information. This is sort of the other side of #796, instead of warning about early iteration, we warn about late mutation.

What I was thinking of was more on the prevention/informational side of things to get an idea of how many holes we need to plug.  A general "you're accessing project state without a lock" warning might be a better approach.

> If not, then are we implying that project model mutation should never happen at execution time?

I think this depends a little bit on what you consider the project model.  In general, I think the answer is "Yes, eventually, but not now" if we are talking about the usual things.  It doesn't make sense to create new extensions, new tasks, new configurations, new variants, etc once we start executing things.  But does it make sense to realize the path to an output file?  Absolutely, and this happens.

**Comment by adammurdoch** [comment]
We shouldn't prevent _realization_ during execution, there are many reasons why you might need to query something about a task that isn't being executed (for example, any kind of export to an IDE or other build system, or reporting), but we might lock down _registration_ using `register()` and _creation_ using `create()` after the task graph has been calculated.
